### PR TITLE
First API Test for the the Frontend for Backends Desktop microservice (webagg)

### PR DESCRIPTION
**Pull request by augustocristian**
Includes:
- Base Class that manages authentication using OAuth2 with password-based authentication to obtain the token.
- Test case utilizing the aforementioned token to access various endpoints of the webagg microservice.
- Consolidated BaseAPIClass and BaseLogged to use and declare same variables
- Cleaned and improved the code, solved several smells reported by Sonarlint+IntelliJ static analysis

### CODE REVIEW

**Comment by augustocristian** [comment]
This PR has the same problem  than #56, I am working on it but in the separate branch: https://github.com/giis-uniovi/retorch-st-eShopContainers/tree/ft_flakyTJobc

**Comment by javiertuya** [comment]
@augustocristian Hold this PR until #56 is solved to do not add new features while the system is not stable

**Comment by augustocristian** [comment]
Now that the SUT is stable, it might be a good idea to continue with this branch, @javiertuya, and finish implementing the API and the end-to-end test cases?

**Comment by javiertuya** [comment]
@augustocristian Maybe, at least, finish this PR to avoid stalled PRs

File: src/test/java/giis/eshopcontainers/e2e/functional/common/BaseAPIClass.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+
+public class BaseAPIClass {
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/common/BaseAPIClass.java:24]: I miss some comment to reflect the nature of this class.
Check consistency between web tests and api tests.

File: src/test/java/giis/eshopcontainers/e2e/functional/common/BaseAPIClass.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        httppost.addHeader("content-type", "application/x-www-form-urlencoded");
+
+        // Add parameters for authentication
+        List<BasicNameValuePair> params = new ArrayList<>(3);
+        params.add(new BasicNameValuePair("grant_type", "password"));
+        params.add(new BasicNameValuePair("client_id", "testalice"));
+        params.add(new BasicNameValuePair("username", "alice"));
+        params.add(new BasicNameValuePair("password", "Pass123$"));
+        params.add(new BasicNameValuePair("client_secret", "secret"));
+        params.add(new BasicNameValuePair("scope", "basket"));
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/common/BaseAPIClass.java:88]: Hardcoded data should be in a properties file. Check consistency between web and api tests

File: src/test/java/giis/eshopcontainers/e2e/functional/common/BaseAPIClass.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+        // Set parameters for HTTP request
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/common/BaseAPIClass.java:90]: This and below: Every statement has tree lines: a blank, a trivial comment and the line...

File: src/test/java/giis/eshopcontainers/e2e/functional/model/Order.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
+package giis.eshopcontainers.e2e.functional.model;
+import java.util.List;
+
+
+
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/model/Order.java:5]: Again extra blank spaces!!

File: src/test/java/giis/eshopcontainers/e2e/functional/tests/WebAggAPITests.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.util.List;
+
+public class WebAggAPITests extends BaseAPIClass {
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/tests/WebAggAPITests.java:27]: I dont get the meaning of WebAgg, could you chose a better name? and some comment here would be useful.

File: src/test/java/giis/eshopcontainers/e2e/functional/tests/WebAggAPITests.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    @Resource(resID = "identity-api", replaceable = {})
+    @AccessMode(resID = "identity-api", concurrency = 50, sharing = true, accessMode = "READONLY")
+    @Resource(resID = "basket-api", replaceable = {})
+    @AccessMode(resID = "basket-api", concurrency = 30,sharing = true, accessMode = "READWRITE")
+    @Resource(resID = "eshopUser", replaceable = {})
+    @AccessMode(resID = "eshopUser", concurrency = 1, accessMode = "READWRITE")
+    @Test
+    @DisplayName("testAddProductsBasketWebAgg")
+    void testAddProductsBasket() throws IOException, InterruptedException {
+        // Initialize Gson for JSON serialization/deserialization
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/tests/WebAggAPITests.java:39]: Trivial comment?

File: src/test/java/giis/eshopcontainers/e2e/functional/tests/WebAggAPITests.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        Gson gson = new Gson();
+
+        addItemsToBasket();
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/tests/WebAggAPITests.java:42]: What is this? the test data setup? the test execution? Here is where a comment would be useful

File: src/test/java/giis/eshopcontainers/e2e/functional/tests/WebAggAPITests.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        Assertions.assertEquals(2, listitems.get(0).getUnits());
+        Assertions.assertEquals(8.5, listitems.get(0).getUnitPrice());
+
+        Assertions.assertEquals(3, listitems.get(1).getProductId());
+        Assertions.assertEquals("Prism White T-Shirt", listitems.get(1).getProductName());
+        Assertions.assertEquals(1, listitems.get(1).getUnits());
+        Assertions.assertEquals(12, listitems.get(1).getUnitPrice());
+    }
+
+    // Method to add items to the basket
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/tests/WebAggAPITests.java:65]: Trivial comment

File: src/test/java/giis/eshopcontainers/e2e/functional/tests/WebAggAPITests.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    public String addItemsToBasket() throws IOException {
+        // Debug log for creating connection
+        log.debug("Creating the connection with URL: http://basket_api_" + tJobName + ":80/api/v1/Basket");
+        // Create HTTP client and POST request
+        HttpClient httpclient = HttpClients.createDefault();
+        String requestURL ="http://webshoppingagg_" + tJobName + ":80/api/v1/Basket";
+        //requestURL = "http://localhost:5016//api/v1/Basket";
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/tests/WebAggAPITests.java:72]: Remove

File: src/test/java/giis/eshopcontainers/e2e/functional/tests/WebAggAPITests.java [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+        StringEntity entity = new StringEntity(json);
+        httpPost.setEntity(entity);
+
+        // Define response handler
+        ResponseHandler<String> responseHandler = new BasicResponseHandler();
+        // Debug log for performing the request
+        log.debug("Performing the request");
+        // Execute HTTP request and return response
+        return httpclient.execute(httpPost, responseHandler);
+    }
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/tests/WebAggAPITests.java:107]: This method and the next have the same problem as the base class, there are many trivial comments

**Review by javiertuya — CHANGES_REQUESTED** [review]

File: src/test/java/giis/eshopcontainers/e2e/functional/model/Order.java [hunk 11]
```diff
@@ -1,9 +1,9 @@
 package giis.eshopcontainers.e2e.functional.model;
-import java.util.List;
-
 
+import java.util.List;
 
 public class Order {
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/model/Order.java:5]: The review didn't include restructuring the models. Get them back to their state before the review.

NOTE: Never add unrelated changes after a revision because this makes extremely difficult the review process. It is good to improve the code, but this must be handled in a separate PR.

**Review by javiertuya — CHANGES_REQUESTED** [review]

File: src/test/java/giis/eshopcontainers/e2e/functional/model/Order.java (hunk 11 above)
Comment (javiertuya) [reply src/test/java/giis/eshopcontainers/e2e/functional/model/Order.java:5]: I don't see the changes, the only change since last review is this:
image

File: src/test/java/giis/eshopcontainers/e2e/functional/model/Order.java
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/model/Order.java:1]: Models have not been reverted

**Review by javiertuya — CHANGES_REQUESTED** [review]

File: src/test/java/giis/eshopcontainers/e2e/functional/model/Order.java (hunk 11 above)
Comment (augustocristian) [reply src/test/java/giis/eshopcontainers/e2e/functional/model/Order.java]: For some reason my IDE reverted to the wrong commit (the previous one instead the original). Sorry for the inconveniences, now the model files are in the original state (the first commit)

File: src/test/java/giis/eshopcontainers/e2e/functional/model/OrderItem.java [hunk 12]
```diff
@@ -11,36 +11,49 @@ public class OrderItem {
...
     }
+
+    public String getPictureUrl() {
+        return pictureUrl;
+    }
+
     public void setPictureUrl(String pictureUrl) {
         this.pictureUrl = pictureUrl;
     }
+// Agregar getters y setters
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/model/OrderItem.java:58]: Why this comment?

**Review by javiertuya — CHANGES_REQUESTED** [review]
Additionally, this branch is behind main

**Comment by augustocristian** [comment]
Branch updated with the lastest combined-dependency update. Removed comments of the restored code, and removed blankspaces in the header and end of the class. Test passing and ready to merge @javiertuya
